flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaTransactionLog.java

```diff
@@ -0,0 +1,307 @@
+                .append(offset)
+                .toString();
+    }
+
+    private Consumer<ConsumerRecord<byte[], byte[]>> maybeAddTransaction(
+            Set<String> openTransactions) {
+        return record -> {
+            final ByteBuffer keyBuffer = ByteBuffer.wrap(record.key());
+            checkKafkaSchemaVersionMatches(keyBuffer);
+            // Ignore 2 bytes because Kafka's internal representation
+            keyBuffer.getShort();
+            final String transactionalId = StandardCharsets.US_ASCII.decode(keyBuffer).toString();
+
+            final ByteBuffer valueBuffer = ByteBuffer.wrap(record.value());
+            checkKafkaSchemaVersionMatches(valueBuffer);
+            final TransactionState state =
+                    TransactionState.fromByte(readTransactionState(valueBuffer));
+
+            LOG.debug("Transaction {} is in state {}", transactionalId, state);
+
+            // If the transaction is in a final state ignore it
+            if (isTransactionInFinalState(state)) {
+                openTransactions.remove(transactionalId);
+                return;
+            }
+
+            final Optional<KafkaWriterState> openTransactionOpt =
+                    fromTransactionalId(transactionalId);
+            // If the transactionalId does not follow the format ignore it
+            if (!openTransactionOpt.isPresent()) {
+                return;
+            }
+
+            final KafkaWriterState openTransaction = openTransactionOpt.get();
+            // If the transactionalId prefixes differ ignore
+            if (!openTransaction
+                    .getTransactionalIdPrefix()
+                    .equals(main.getTransactionalIdPrefix())) {
+                LOG.debug(
+                        "The transactionalId prefixes differ. Open: {}, Recovered: {}",
+                        openTransaction.getTransactionalIdPrefix(),
+                        main.getTransactionalIdPrefix());
+                return;
+            }
+
+            final int openSubtaskIndex = openTransaction.getSubtaskId();
+            final long openCheckpointOffset = openTransaction.getTransactionalIdOffset();
+
+            final boolean commonRemainder =
+                    openSubtaskIndex % numberOfParallelSubtasks == main.getSubtaskId()
+                            && openSubtaskIndex != main.getSubtaskId();
+            final boolean sameSubtaskHigherCheckpoint =
+                    hasSameSubtaskWithHigherCheckpoint(openSubtaskIndex, openCheckpointOffset);
+            final boolean higherSubtasksWithFirstCheckpoint =
+                    minOthers.isPresent()
+                            && minOthers.getAsInt() < openSubtaskIndex
+                            && openCheckpointOffset == 0;
+
+            if (!commonRemainder
+                    && !sameSubtaskHigherCheckpoint
+                    && !higherSubtasksWithFirstCheckpoint) {
+                return;
+            }
```

Comment:
Please extract that logic into a test method and add unit tests + comments.

Reply:
Just to be sure: this procedure would also abort all transactions from "unknown" subtasks because the checkpoint would not reflect the true previous parallelism (upscale-downscale without checkpoint), correct?